server/docker_resource.py

```diff
@@ -0,0 +1,343 @@
+            raise ValidationException('Docker images were not proper')
+
+        event.preventDefault().stopPropagation()
+
+
+# TODO check local docker cache and default registry
+def imageExists(name):
+    """Given an image name determine if the image exists locally or on the default
+    registry"""
+
+    return True
+
+
+def checkOldImage(key, dataDict):
+    localImage = localDockerImageExists(dataDict["name"])
+    if localImage and localImage != key:
+        print('docker cache does not match the version in girder')
+        print ('compare keys:\n,%s\n,%s' % key, localImage)
+
+
+def checkNewImage(name,cache):
+    foundImage = False
+    for (dictKey, dictValue) in iteritems(cache):
+        if dictValue and "name" in dictValue:
+            if name == dictValue["name"]:
+                foundImage=True
+    if foundImage or not imageExists(name):
+        return False
+    else:
+        return True
+
+
+def localDockerImageExists(imageName):
+    """checks the local docker cache for the image
+    :param imageName: the docker image name in the form of repo/name:tag
+    if the tag is not given docker defaults to using the :latest tag
+    :type imageName: string
+    :returns: if the image exit the id(sha256 hash) is returned otherwise
+    None is returned
+    """
+    try:
+        # docker inspect returns non zero if the image is not available
+        # locally
+        data = subprocess.checkoutput('docker', 'inspect',
+                                      '--format="{{json .Id}}"', imageName)
+    except subprocess.CalledProcessError as err:
+        # the image does not exist locally, try to pull from dockerhub
+
+        return None
+
+    return data
+
+
+def getDockerImages():
+    module_list = ModelImporter.model('setting').get(
+        PluginSettings.DOCKER_IMAGES)
+    if module_list is None:
+        module_list = {}
+    return module_list
+
+
+class DockerCache() :
+    imageName='docker_image_name'
+    type='type'
+    xml='xml'
+
+    def __init__(self, cache):
+        """
+        Data is stored in the following format:
+            {image_name_hash:
+                {
+                cli_name:{
+                    type:
+                    xml:
+
+                    }
+                docker_image_name:name
+                }
+            }
+        """
+
+        self.data = cache
+
+    def getDockerImg(self):
+        nameList = []
+        for (imgHash, imgDict) in iteritems(self.data):
+            nameList.append(str(imgDict[self.imageName]))
+
+        return nameList
+
+    def imageAlreadyLoaded(self, name):
+        imageKey = self._getHashKey(name)
+        if imageKey in self.data:
+            return True
+        else:
+            return False
+
+    def addImage(self, name):
+        imageKey = self._getHashKey(name)
+        self.data[imageKey] = {}
+
+        self.data[imageKey][self.imageName] = name
+
+    def addCLI(self, img_name, cli_name, type, xml):
```

Comment:
This makes me think that it would be nice have another class called `DockerImage` which manages a list of CLIs with methods to addCLI, deleteCLI, getCLIXML etc. And DockerCache manages a list of `DockerImage` objects with methods to `addDockerImage, deleteDockerImage, updateDockerImage` etc.